plugins/in_podman_metrics/podman_metrics.c

```diff
@@ -0,0 +1,496 @@
+        flb_plg_warn(ctx->ins, "Failed to parse JSON %d: %s", r, buffer);
+        return -1;
+    }
+
+    flb_plg_debug(ctx->ins, "Got %d nested tokens", t[0].size);
+
+    if (r < 1 || t[0].type != JSMN_ARRAY) {
+        flb_plg_warn(ctx->ins, "Expected array at the json root");
+        return -1;
+    }
+
+    for (i=0; i<r; i++) {
+        if (t[i].type == JSMN_STRING) {
+            if ((int)strlen(JSON_FIELD_ID) == t[i].end - t[i].start &&
+                strncmp(buffer + t[i].start, JSON_FIELD_ID, t[i].end - t[i].start) == 0) {
+                memset(id, 0, sizeof(id));
+                strncpy(id, buffer + t[i+1].start, t[i + 1].end - t[i + 1].start);
+                id[strlen(id)] = '\0';
+                flb_plg_trace(ctx->ins, "Found id %s", id);
+            }
+            else if ((int)strlen(JSON_FIELD_NAMES) == t[i].end - t[i].start &&
+                     strncmp(buffer + t[i].start, JSON_FIELD_NAMES, t[i].end - t[i].start) == 0) {
+                array_id = i + 1;
+                if (t[array_id].type == JSMN_ARRAY)
+                {
+                    j = array_id + 1;
+                    while (t[j].parent == array_id)
+                    {
+                        memset(name, 0, sizeof(name));
+                        strncpy(name, buffer + t[j].start, t[j].end - t[j].start);
+                        name[strlen(name)] = '\0';
+                        flb_plg_trace(ctx->ins, "Found name %s", name);
+                        add_container_to_list(ctx, id, name);
+                        j++;
+                        collected_containers++;
+                    }
+                }
+            }
+        }
+    }
+
+    flb_plg_debug(ctx->ins, "Collected %d containers from podman config file", collected_containers);
+    return collected_containers;
+}
+
+/*
+ * Create structure instance based on previously found id and name. Set all its values (like
+ * memory or cpu to UINT64_MAX, in case it won't be found later. This function also adds this structure
+ * to internal list, so it can be found by iteration later on.
+ */
+static int add_container_to_list(struct flb_in_metrics *ctx, flb_sds_t id, flb_sds_t name)
+{
+    struct container *cnt;
+    cnt = flb_malloc(sizeof(struct container));
+    cnt->id = flb_sds_create(id);
+    cnt->name = flb_sds_create(name);
+
+    cnt->memory_usage = UINT64_MAX;
+    cnt->memory_max_usage = UINT64_MAX;
+    cnt->memory_limit = UINT64_MAX;
+    cnt->rss = UINT64_MAX;
+    cnt->cpu_user = UINT64_MAX;
+    cnt->cpu = UINT64_MAX;
+
+    mk_list_init(&cnt->net_data);
+
+    mk_list_add(&cnt->_head, &ctx->items);
+    return 0;
+}
+
+/*
+ * Iterate over container list and remove collected data. Also call
+ * destroy_counter() on every created counter.
+ */
+static int destroy_container_list(struct flb_in_metrics *ctx)
+{
+    struct container *cnt;
+    struct net_iface *iface;
+    struct sysfs_path *pth;
+    mk_list_foreach_safe(ctx->head, ctx->tmp, &ctx->items) {
```

Comment:
it's not necessary to have ctx->`struct mk_list tmp` or cnt->`struct mk_list tmp`, just declare those list iterators in the function that you need them, not in the contexts.